### ISSUE TITLE
Get error when i use multithreading...

### ISSUE THREAD

**Issue by mathieu-genete** [itsmeludo/PhylOligo#8]
Hi,

I've recently installed phyloligo on our server to test it. It perfectly work only if i don't use multithreading.
When i want to use more then 1 cpu (see command line bellow), i get an error (see bellow)

Thank's for your help...

Mathieu

we have python 3.4 installed.

command:_phyloligo.py -i trinity/E3/Trinity.fasta -d JSD -o phyloligo_results/E3_JSD_v2.mat --method joblib -c 16_

error:
_sklearn.externals.joblib.externals.loky.process_executor._RemoteTraceback: 
'''
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/externals/loky/process_executor.py", line 393, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/usr/lib/python3.4/multiprocessing/queues.py", line 115, in get
    return ForkingPickler.loads(res)
AttributeError: Can't get attribute 'JSD' on <module 'sklearn.externals.joblib.externals.loky.backend.popen_loky_posix' from '/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/externals/loky/backend/popen_loky_posix.py'>
'''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/gepv/home2/tools/PhylOligo-master/src/phyloligo.py", line 1155, in <module>
    ret = main()      
  File "/gepv/home2/tools/PhylOligo-master/src/phyloligo.py", line 1139, in main
    params.dist, params.threads_max, params.freqchunksize, params.workdir)
  File "/gepv/home2/tools/PhylOligo-master/src/phyloligo.py", line 631, in compute_distances
    res = compute_distances_joblib(frequencies, metric=dist, n_jobs=threads_max)
  File "/gepv/home2/tools/PhylOligo-master/src/phyloligo.py", line 473, in compute_distances_joblib
    distances = pairwise_distances(frequencies, metric=call_dist[metric], n_jobs=n_jobs)
  File "/usr/local/lib/python3.4/dist-packages/sklearn/metrics/pairwise.py", line 1402, in pairwise_distances
    return _parallel_pairwise(X, Y, func, n_jobs, **kwds)
  File "/usr/local/lib/python3.4/dist-packages/sklearn/metrics/pairwise.py", line 1075, in _parallel_pairwise
    for s in gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs)))
  File "/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/parallel.py", line 996, in __call__
    self.retrieve()
  File "/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/parallel.py", line 899, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/_parallel_backends.py", line 517, in wrap_future_result
    return future.result(timeout=timeout)
  File "/usr/lib/python3.4/concurrent/futures/_base.py", line 402, in result
    return self.__get_result()
  File "/usr/lib/python3.4/concurrent/futures/_base.py", line 354, in __get_result
    raise self._exception
sklearn.externals.joblib.externals.loky.process_executor.BrokenProcessPool: A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable._

**Comment by T-B-F** [comment]
Hi, and thank you for the report.
Could you give me the version of your sklearn package?

And try this gist:
https://gist.github.com/T-B-F/11b9527687d07eb903614fb59455073c

**Comment by mathieu-genete** [comment]
Hi,
Thanks for your response. Our sklearn package version is 0.20.0.
I will test the gist and will give you a feedback.

**Comment by mathieu-genete** [comment]
I've launch the "parallel_jsd_joblib.py" file, i've the same error:

0.20.0
1.15.2
sklearn.externals.joblib.externals.loky.process_executor._RemoteTraceback: 
'''
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/externals/loky/process_executor.py", line 393, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/usr/lib/python3.4/multiprocessing/queues.py", line 115, in get
    return ForkingPickler.loads(res)
AttributeError: Can't get attribute 'JSD' on <module 'sklearn.externals.joblib.externals.loky.backend.popen_loky_posix' from '/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/externals/loky/backend/popen_loky_posix.py'>
'''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "parallel_jsd_joblib.py", line 65, in <module>
    distances = pairwise_distances(frequencies, metric=JSD, n_jobs=2)
  File "/usr/local/lib/python3.4/dist-packages/sklearn/metrics/pairwise.py", line 1402, in pairwise_distances
    return _parallel_pairwise(X, Y, func, n_jobs, **kwds)
  File "/usr/local/lib/python3.4/dist-packages/sklearn/metrics/pairwise.py", line 1075, in _parallel_pairwise
    for s in gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs)))
  File "/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/parallel.py", line 996, in __call__
    self.retrieve()
  File "/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/parallel.py", line 899, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/usr/local/lib/python3.4/dist-packages/sklearn/externals/joblib/_parallel_backends.py", line 517, in wrap_future_result
    return future.result(timeout=timeout)
  File "/usr/lib/python3.4/concurrent/futures/_base.py", line 402, in result
    return self.__get_result()
  File "/usr/lib/python3.4/concurrent/futures/_base.py", line 354, in __get_result
    raise self._exception
sklearn.externals.joblib.externals.loky.process_executor.BrokenProcessPool: A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.

**Comment by T-B-F** [comment]
Ok, it seems to be a sklearn/joblib problem with their new loky backend.
Could you try to run the gist script with sklearn 0.19.1?

I will see if there is a workaround this loky stuff.

**Comment by mathieu-genete** [comment]
Ok,

I found a solution that works well.
I've this output:
0.20.0
1.15.2
(100, 100)

I modified code as this:

**file parallel_jsd_joblib.py:**
```
from sklearn.metrics.pairwise import pairwise_distances
import numpy as np
import sklearn
import defs

print(sklearn.__version__)
print(np.__version__)

a = np.random.random(10)
b = np.random.random(10)

frequencies = np.random.random((100, 10))
distances = pairwise_distances(frequencies, metric=defs.JSD, n_jobs=2)
print(distances.shape)
```

**I created a file with functions:**

defs.py:

```
import numpy as np

def posdef_check_value(d):
    d[np.isnan(d)]=0
    d[np.isinf(d)]=0

## Distance functions                                                                                                                                                                                                                                                           

def KL(a, b):
    """ compute the KL distance                                                                                                                                                                                                                                                 
    """
    if a.ndim == 1 and b.ndim == 1:
	d = a * np.log(a/b)
	posdef_check_value(d)
	res = np.sum(d)
    elif a.ndim == 2 and b.ndim == 2:
	X, Y = check_pairwise_arrays(a, b)
	X = X[:,np.newaxis]
	d = X * np.log(X/Y)
	posdef_check_value(d)
	res = np.sum(d, axis=2).T
    else:
        print("Dimension erro in KL, a={}, b={}".format(a.ndim, b.ndim), file=sys.stderr)
	sys.exit(1)
    return res

def JSD(a, b):
    """ Compute JSD distance                                                                                                                                                                                                                                                    
    """
    if a.ndim == 1 and b.ndim == 1:
	h = 0.5 * (a + b)
	d = 0.5 * (KL(a, h) + KL(b, h))
    elif a.ndim==2 and b.ndim == 1:
	h = 0.5 * (a[np.newaxis,:] + b)
	d1 = a[np.newaxis,:] * np.log(a[np.newaxis,:]/h)
	posdef_check_value(d1)
	d1 = np.sum(d1, axis=2)
	d2 = b * np.log(b/h)
	posdef_check_value(d2)
	d2 = np.sum(d2, axis=2)
	d = 0.5 * (d1 + d2)
    else:
        h = 0.5 * (a[np.newaxis,:] + b[:, np.newaxis])
	d1 = a[np.newaxis,:] * np.log(a[np.newaxis,:]/h)
	posdef_check_value(d1)
	d1 = np.sum(d1, axis=2)
	d2 = b[:, np.newaxis] * np.log(b[:, np.newaxis]/h)
	posdef_check_value(d2)
	d2 = np.sum(d2, axis=2)
	d = 0.5 * (d1 + d2)
	#d = d.T                                                                                                                                                                                                                                                                
    return d
```

**Comment by mathieu-genete** [comment]
source: https://stackoverflow.com/questions/41385708/multiprocessing-example-giving-attributeerror

**Comment by T-B-F** [comment]
Great and many thanks! I never had and never heard about this issue before (fixed d93b02b)
